### PR TITLE
modpcap reader support

### PR DESCRIPTION
**Pull request by oosthoek**
support for modified pcaps
https://wiki.wireshark.org/Development/LibpcapFileFormat#modified-pcap

### CODE REVIEW

File: dpkt/pcap.py [hunk 1]
```diff
@@ -158,10 +163,52 @@ class PktHdr(dpkt.Packet):
...
+    __hdr__ = (
+        ('pkt_hdr_buf', f"{PktHdr.__hdr_len__}s", PktHdr.__hdr_len__ * b'\x00'),
+        ('ifindex', 'I', 0),
+        ('protocol', 'H', 0),
+        ('pkt_type', 'B', 0),
+        ('pad', 'B', 0),
+    )
+
+    def get_hdr(self):
+        return PktHdr(self.pkt_hdr_buf)
```
Comment (obormot) [inline dpkt/pcap.py:189]: why wouldn't this return the full mod header?

File: dpkt/pcap.py [hunk 2]
```diff
@@ -277,17 +326,24 @@ def __init__(self, fileobj):
...
+
+        self.__ph = MAGIC_TO_PKT_HDR[magic]
+
+
         if self.__fh.linktype in dltoff:
             self.dloff = dltoff[self.__fh.linktype]
         else:
             self.dloff = 0
-        self._divisor = 1E6 if self.__fh.magic in (TCPDUMP_MAGIC, PMUDPCT_MAGIC) else Decimal('1E9')
+        self._divisor = 1E6 if magic in (TCPDUMP_MAGIC, PMUDPCT_MAGIC) else Decimal('1E9')
```
Comment (obormot) [inline dpkt/pcap.py:346]: the new magic should be added here I think, or better reverse the logic:
```
self._divisor = Decimal('1E9') if magic in (TCPDUMP_MAGIC_NANO, PMUDPCT_MAGIC_NANO) else 1E6
```

File: dpkt/pcap.py [hunk 3]
```diff
@@ -158,10 +163,52 @@ class PktHdr(dpkt.Packet):
...
     __byte_order__ = '<'
 
+class LEPktModHdr(PktModHdr):
+    __byte_order__ = '<'
+
+    def get_hdr(self):
+        return LEPktHdr(self.pkt_hdr_buf)
+
+
+
```
Comment (obormot) [inline dpkt/pcap.py:203]: extra whitespace

File: dpkt/pcap.py [hunk 4]
```diff
@@ -187,6 +234,8 @@ class LEFileHdr(FileHdr):
     __byte_order__ = '<'
 
 
+
+
```
Comment (obormot) [inline dpkt/pcap.py:238]: extra whitespace

File: dpkt/pcap.py [hunk 5]
```diff
@@ -277,17 +326,24 @@ def __init__(self, fileobj):
...
             self.__fh = LEFileHdr(buf)
-            self.__ph = LEPktHdr
-        elif self.__fh.magic not in (TCPDUMP_MAGIC, TCPDUMP_MAGIC_NANO):
+
+        if magic not in MAGIC_TO_PKT_HDR:
             raise ValueError('invalid tcpdump header')
+
+        self.__ph = MAGIC_TO_PKT_HDR[magic]
+
+
```
Comment (obormot) [inline dpkt/pcap.py:341]: extra whitespace

**Review by obormot — CHANGES_REQUESTED** [review]

File: dpkt/pcap.py [hunk 6]
```diff
@@ -519,6 +582,21 @@ def test_reader_fd():
         assert reader.fd == fd.fileno()
         assert reader.fileno() == fd.fileno()
 
+def test_reader_modified_pcap_type():
```
Comment (obormot) [inline dpkt/pcap.py:585]: needs 1 extra whitespace before function def

File: dpkt/pcap.py [hunk 7]
```diff
@@ -158,10 +163,52 @@ class PktHdr(dpkt.Packet):
...
+    https://wiki.wireshark.org/Development/LibpcapFileFormat#modified-pcap
+
+    TODO: Longer class information....
+
+    Attributes:
+        __hdr__: Header fields of pcap header.
+        TODO.
+    """
+    __hdr__ = (
+        ('pkt_hdr_buf', f"{PktHdr.__hdr_len__}s", PktHdr.__hdr_len__ * b'\x00'),
```
Comment (obormot) [inline dpkt/pcap.py:181]: f-strings aren't supported by all versions of python in dpkt's build pipeline I think

File: dpkt/pcap.py (hunk 1 above)
Comment (obormot) [reply dpkt/pcap.py:189]: yeah I'd rather not add the `get_hdr()` method to the existing `PktHdr` class for this reason alone, I don't think this is justified.  I'd just duplicate the orig fields in the new header
```
class PktModHdr(dpkt.Packet):
    ...
    __hdr__ = (
        ('tv_sec', 'I', 0),
        ('tv_usec', 'I', 0),
        ('caplen', 'I', 0),
        ('len', 'I', 0),
        ('ifindex', 'I', 0),
        ('protocol', 'H', 0),
        ('pkt_type', 'B', 0),
        ('pad', 'B', 0),
    )
```
this reads cleaner and doesn't change the existing code.

**Review by obormot — CHANGES_REQUESTED** [review]

**Review by obormot — APPROVED** [review]

**Comment by obormot** [comment]
Thanks for your contribution @oosthoek !

**Comment by oosthoek** [comment]
@obormot thanks!  is there a pypi release planned anytime soon?

**Comment by obormot** [comment]
Don't have it planned, but I've just created https://github.com/kbandla/dpkt/issues/634 to get it started when @brifordwylie or I get time to dedicate. Hopefully soon!

**Pull request merged**